mesonbuild/scripts/run_tool.py

```diff
@@ -3,17 +3,86 @@
 
 from __future__ import annotations
 
-import itertools
+import asyncio
+import asyncio.subprocess
 import fnmatch
-import concurrent.futures
+import itertools
+import signal
+import sys
 from pathlib import Path
 
+from .. import mlog
 from ..compilers import lang_suffixes
 from ..mesonlib import quiet_git
+from ..mtest import sh_quote, determine_worker_count, complete_all
 import typing as T
 
-if T.TYPE_CHECKING:
-    import subprocess
+Info = T.TypeVar("Info")
+
+async def run_with_buffered_output(cmdlist: T.List[str]) -> int:
+    """Run the command in cmdlist, buffering the output so that it is
+       not mixed for multiple child processes.  Kill the child on
+       cancellation."""
+    quoted_cmdline = ' '.join(sh_quote(x) for x in cmdlist)
+    p: T.Optional[asyncio.subprocess.Process] = None
+    try:
+        p = await asyncio.create_subprocess_exec(*cmdlist,
+                                                 stdin=asyncio.subprocess.DEVNULL,
+                                                 stdout=asyncio.subprocess.PIPE,
+                                                 stderr=asyncio.subprocess.STDOUT)
```

Comment:
I think that this means we kill colored output since tools being run are not going to see isatty() return true anymore... this is a bit unfortunate.

Reply:
Hmm yeah, it's like ninja in this respect but it's unfortunate for clang-tidy. I will make it respect b_colorout=always/auto.

Reply:
Since this uses the console pool, ninja directly connects run_tool consumers such as `ninja clippy clang-format clang-tidy` to the tty, and os.isatty(sys.stdout) returns true. `ninja clang-format-check` etc. will actually produce colored statuses (albeit potentially interleaved a bit randomly).

The question is how to use b_colorout. Would that mean just to disable the asyncio pool and disable stdout pipe buffering? Is it possible to conveniently fake `isatty()` the way e.g. /usr/bin/script does?

Reply:
Yeah, I meant "it's like ninja" in the sense that for compilers you need --color=always due to ninja's buffering (which is what I'm trying to mimic here).

> The question is how to use b_colorout

Since we know what tools we're running, we can add --color=always|never options by hand in mesonbuild/scripts. For clang-tidy that's actually a bugfix in the opposite case: `ninja clang-tidy` was not respecting -Db_colorout=never.

For clippy and (in the future) rustfmt we get the `--color=always|auto|never option` for free from the rustc command line.

> Is it possible to conveniently fake isatty() the way e.g. /usr/bin/script does?

script uses a pty which is pretty hard to do, especially for Windows. We can fake it by using the isatty value inside the script when -Db_colorout=auto, i.e. using isatty to change auto to `always` or `never`.

Reply:
Basically what I mean to say here is that I have some kind of recollection that I tried to fix interleaved output here a couple years ago and threw up my hands in disgust upon learning that clang tooling didn't actually support command line options to enable color anyways.

Maybe it's changed but on the other hand it looks like your latest push only handles tidy?

Reply:
yeah, but I found out clang-format also supports -f{,no-}color-diagnostics so it can be fixed as well

Reply:
Could be informative to check which version of clang first added the undocumented support.

Reply:
If it's since [https://github.com/llvm/llvm-project/commit/692175399453d70083772033e13a11e408576327](this) it would be LLVM 7.0.0. I'll try to find some old distros, but it may not be this week.